jax/_src/random.py

```diff
@@ -1562,32 +1569,46 @@ def categorical(key: ArrayLike,
     shape: Optional, a tuple of nonnegative integers representing the result shape.
       Must be broadcast-compatible with ``np.delete(logits.shape, axis)``.
       The default (None) produces a result shape equal to ``np.delete(logits.shape, axis)``.
+    replace: Perform sampling without replacement.
 
   Returns:
     A random array with int dtype and shape given by ``shape`` if ``shape``
     is not None, or else ``np.delete(logits.shape, axis)``.
+
+  References:
+    .. [1] Wouter Kool, Herke van Hoof, Max Welling. "Stochastic Beams and Where to Find
+      Them: The Gumbel-Top-k Trick for Sampling Sequences Without Replacement".
+      Proceedings of the 36th International Conference on Machine Learning, PMLR
+      97:3499-3508, 2019. https://proceedings.mlr.press/v97/kool19a.html.
   """
   key, _ = _check_prng_key("categorical", key)
   check_arraylike("categorical", logits)
-  logits_arr = jnp.asarray(logits)
-
-  if axis >= 0:
-    axis -= len(logits_arr.shape)
-
-  batch_shape = tuple(np.delete(logits_arr.shape, axis))
+  logits = jnp.asarray(logits)
+  axis = canonicalize_axis(axis, logits.ndim)
   if shape is None:
-    shape = batch_shape
+    shape = util.tuple_delete(logits.shape, axis)
   else:
     shape = core.canonicalize_shape(shape)
-    _check_shape("categorical", shape, batch_shape)
```

Comment:
We should not remove this check – it validates that `shape` meets the documented requirements.